### PR TITLE
MIDI Synth - Big Update

### PR DESCRIPTION
**Pull request by dylwhich**
### Description

**MIDI Changes**
* ADSR!
* Multiple instrument banks
* Simple pitch-shifted samples
* Actual instrument definitions without hardcoded behavior
* Many small things
* platform-midi library submodule for basic OS MIDI support

**Synth Mode Changes**
* New instrument icons
* Lots of menus for controllers, etc.
* Added wheel menu and related icons

**Font Stuff**
* drawTextBounds(): Lets you write text, chopped off at arbitrary bounds
* drawTextMarquee(): It's back, and it's coming to your swadge!
* drawTextEllipsize(): Auto-ellipsizes text within a boundary
* drawTextMulticolored(): Lets you draw rainbow-y text and such
* Fixed an oversight in drawTextWordWrap() that made it un-continuable

**Drawing Functions**
* drawRectFilled(), because fillDisplayArea() is much more fragile
* drawCircleFilledQuadrants() to go along with drawCircleQuadrants()
* drawRoundedRect()

**Menu Stuff**
* Mania menu renderer uses drawTextEllipsize() if necessary (or drawTextMarquee() for the currently selected item)
* Menu now updates 'current value' properly for settings option items
* Utilities for easily saving and restoring the menu position between reconstruction

**Other**
* addBefore() / addAfter() functions for linked_list
* All sorts of fixes and improvements for wheel menu
* Fixed a crash in the raw file preprocessor

**Documentation**
* Updated docs for DNF packages which have changed slightly since last time

### Test Instructions

Go crazy on the synth mode in the emulator and the hardware.

### Readiness Checklist
- [x] I have run `make format` to format the changes
- [x] I have compiled the firmware and the changes have no warnings
- [x] I have compiled the emulator and the changes have no warnings
- [x] I have run `make cppcheck` and checked that `cppcheck_result.txt` has no warnings for the changes
- [x] I have added doxygen comments to any code used by more than one Swadge mode. This includes `/*! \file` comments with Design Philosophy, Usage, and Example sections for new headers.
- [x] I have run `make docs` and checked that `doxy_warnings.txt` has no warnings for the new code

### CODE REVIEW

File: main/utils/wheel_menu.c [hunk 1]
```diff
@@ -137,42 +137,64 @@ void wheelMenuSetItemInfo(wheelMenuRenderer_t* renderer, const char* label, cons
...
+void wheelMenuSetItemTextIcon(wheelMenuRenderer_t* renderer, const char* label, const char* textIcon)
 {
-    node_t* node = menu->items->first;
+    wheelItemInfo_t* info = findOrAddInfo(renderer, label);
+
+    info->textIcon = textIcon;
+}
 
+/**
+ * @brief Set specific dimensions for a
```
Comment (AEFeinstein) [inline main/utils/wheel_menu.c:171]: For a what? FOR A WHAT?

**Review by AEFeinstein — APPROVED** [review]
There are a few new warnings which should be cleaned up, but none look like errors in disguise. Make sure to clean, then build to see them all!

**Comment by AEFeinstein** [comment]
Also, being honest, I skimmed this review. It'd have been a lot to read...

**Review by AEFeinstein — APPROVED** [review]

File: emulator/idf-inc/class/midi/midi_device_emu.h [hunk 2]
```diff
@@ -0,0 +1 @@
+void setMidiClientName(const char* name);
```
[the remark this answers is not in the document]
Comment (dylwhich) [reply emulator/idf-inc/class/midi/midi_device_emu.h]: Man how do I keep forgetting this!

File: attic/components/hdw-spiffs/hdw-spiffs.c
[the remark this answers is not in the document]
Comment (dylwhich) [reply attic/components/hdw-spiffs/hdw-spiffs.c:1]: Ah, this was something I had added before the CNFS switchover. The idea was to let you drop additional MIDI files in the `spiffs_image` folder post-compile and be able to play them just like the built-in ones.

File: main/display/font.c [hunk 3]
```diff
@@ -453,3 +507,119 @@ void makeOutlineFont(font_t* srcFont, font_t* dstFont, bool spiRam)
...
+ * @param color The color of the character to draw
+ * @param text  The text to draw to the display
+ * @param xOff  The x offset to draw the text at
+ * @param yOff  The y offset to draw the text at
+ * @param xMax  The x offset of the right edge of the marquee text
+ * @param[in,out] timer A pointer to a timer value used to time the marquee
+ * @return The x offset at the end of the drawn string
+ */
+int16_t drawTextMarquee(const font_t* font, paletteColor_t color, const char* text, int16_t xOff, int16_t yOff,
+                        int16_t xMax, int64_t* timer)
```
[the remark this answers is not in the document]
Comment (dylwhich) [reply main/display/font.c]: Ah, I didn't notice that so many other places were already using 32-bit timers! Changed it everywhere it's currently being used. Hopefully nobody needs to scroll text slower than once every half an hour lol

**Comment by dylwhich** [comment]
> Also, being honest, I skimmed this review. It'd have been a lot to read...

That's pretty fair, it's over 200 commits and like a billion lines. As long as the library code changes look good!

**Comment by AEFeinstein** [comment]
Something something more frequent, smaller PRs please

**Pull request merged**